pkg/controller/defaults/region.go

```diff
@@ -3,30 +3,36 @@ package defaults
 import (
 	"context"
 
-	internalv1 "github.com/acorn-io/acorn/pkg/apis/internal.acorn.io/v1"
 	"github.com/acorn-io/acorn/pkg/labels"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func addDefaultRegion(ctx context.Context, c client.Client, appInstance *internalv1.AppInstance) error {
-	if appInstance.Spec.Region != "" {
-		appInstance.Status.Defaults.Region = ""
-		return nil
-	}
+type RegionGetterSetter interface {
+	metav1.Object
+	GetRegion() string
+	SetDefaultRegion(string)
+}
 
-	ns := new(corev1.Namespace)
-	if err := c.Get(ctx, client.ObjectKey{Name: appInstance.Namespace}, ns); err != nil {
-		return err
-	}
+func AddDefaultRegion(ctx context.Context, c client.Client, obj RegionGetterSetter) error {
+	var region string
+	if obj.GetRegion() == "" {
+		ns := new(corev1.Namespace)
+		if err := c.Get(ctx, client.ObjectKey{Name: obj.GetNamespace()}, ns); err != nil {
+			return err
+		}
 
-	appInstance.Status.Defaults.Region = ns.Annotations[labels.AcornProjectDefaultRegion]
-	if appInstance.Status.Defaults.Region == "" {
-		appInstance.Status.Defaults.Region = ns.Annotations[labels.AcornCalculatedProjectDefaultRegion]
-	}
-	if appInstance.Status.Defaults.Region == "" {
-		appInstance.Status.Defaults.Region = "local"
+		region = ns.Annotations[labels.AcornProjectDefaultRegion]
+		if region == "" {
+			if r := ns.Annotations[labels.AcornCalculatedProjectDefaultRegion]; r != "" {
+				region = r
+			} else {
+				region = "local"
+			}
+		}
 	}
 
+	obj.SetDefaultRegion(region)
```

Comment:
should we move this line above inside the if block so that we only set defaultRegion when obj.GetRegion() is empty?

Reply:
I don't think so. There are situations where we want to change the default region even if it is set. For example: builders. We always want the builder to run in the default region. So, if the default region for a project changes, then this should be updated.

Reply:
Got it. But in this case, should we always get default region from its project. The current logic will set default region to empty string if obj.GetRegion() != ""

Reply:
Yep, you're right. Updated.